src/web-extensions.c

```diff
@@ -0,0 +1,794 @@
+  const char *arg_name;
+  char *arg_extension_or_origin;
+  const char *app_id;
+  g_autofree char *host_path = NULL;
+  g_autofree char *description = NULL;
+  g_autofree char *manifest_filename = NULL;
+  guint response = XDG_DESKTOP_PORTAL_RESPONSE_OTHER;
+  gboolean should_close_session;
+  Permission permission;
+  gboolean allowed;
+  char *argv[] = {NULL, NULL, NULL, NULL};
+  g_autoptr(GError) error = NULL;
+
+  REQUEST_AUTOLOCK (request);
+  session = g_object_get_data (G_OBJECT (request), "session");
+  SESSION_AUTOLOCK_UNREF (g_object_ref (session));
+  g_object_set_data (G_OBJECT (request), "session", NULL);
+  web_extensions_session = (WebExtensionsSession *)session;
+
+  if (!request->exported || web_extensions_session->state != WEB_EXTENSIONS_SESSION_STATE_STARTING)
+    goto out;
+
+  arg_name = g_object_get_data (G_OBJECT (request), "name");
+  arg_extension_or_origin = g_object_get_data (G_OBJECT (request), "extension-or-origin");
+
+  host_path = find_messaging_host (web_extensions_session->mode,
+                                   arg_name, arg_extension_or_origin,
+                                   &description, &manifest_filename, NULL,
+                                   &error);
+  if (host_path == NULL)
+    {
+      g_warning ("Could not find WebExtensions backend: %s", error->message);
+      fflush(stderr);
+      fflush(stdout);
+      goto out;
+    }
+
+  app_id = xdp_app_info_get_id (request->app_info);
+  permission = get_permission_sync (app_id, PERMISSION_TABLE, arg_name);
+  if (permission == PERMISSION_ASK || permission == PERMISSION_UNSET)
+    {
+      guint access_response = 2;
+      g_autoptr(GVariant) access_results = NULL;
+      GVariantBuilder opt_builder;
+      g_autoptr(GAppInfo) info = NULL;
+      const char *display_name;
+      g_autofree gchar *app_info_id = NULL;
+      g_autofree gchar *title = NULL;
+      g_autofree gchar *subtitle = NULL;
+      g_autofree gchar *body = NULL;
+
+      info = xdp_app_info_load_app_info (request->app_info);
+      if (info)
+        {
+          g_auto(GStrv) app_id_components = g_strsplit (g_app_info_get_id (info), ".desktop", 2);
+          app_info_id = g_strdup (app_id_components[0]);
+        }
+      display_name = info ? g_app_info_get_display_name (info) : app_id;
+      title = g_strdup_printf (_("Allow %s to start WebExtension backend?"), display_name);
+      subtitle = g_strdup_printf (_("%s is requesting to launch \"%s\" (%s)."), display_name, description, arg_name);
+      body = g_strdup (_("This permission can be changed at any time from the privacy settings."));
+
+      g_variant_builder_init (&opt_builder, G_VARIANT_TYPE_VARDICT);
+      g_variant_builder_add (&opt_builder, "{sv}", "deny_label", g_variant_new_string (_("Don't allow")));
+      g_variant_builder_add (&opt_builder, "{sv}", "grant_label", g_variant_new_string (_("Allow")));
+      if (!xdp_dbus_impl_access_call_access_dialog_sync (access_impl,
```

Comment:
As things currently stand, `handle_start_in_thread` is holding the session object's mutex when it pops up the access dialog. So a call to the Close method would block when it tried to acquire the mutex. The same would be true if the browser exited and we were trying to clean up.

What does seem to be missing is that we aren't passing the cancellable to `call_access_dialog_sync`. This would let the browser withdraw the access dialog early by calling `Close` on the Request object returned by `Start`, and ensure the cleanup (in `close_requests_in_thread_func`) doesn't block should the browser exit.

Reply:
@bandali0 What was described above is not implemented in the Firefox-side patch. To see the relevant code, use Ctrl+F "closeSession" at https://phabricator.services.mozilla.com/D140803. Are you going to implement it?

On the portal patch here, please add a comment that explains that the responsibility of graceful killing the application has been delegated to Firefox, and that the SIGKILL here is a last shot attempt to clean up if needed.

Reply:
> What does seem to be missing is that we aren't passing the cancellable to `call_access_dialog_sync`.

The cancellable is now passed in https://github.com/flatpak/xdg-desktop-portal/pull/705/commits/c73983517ddc90a023e0b308cc5e8abeff496f5a .

> This would let the browser withdraw the access dialog early by calling `Close` on the Request object returned by `Start`, and ensure the cleanup (in `close_requests_in_thread_func`) doesn't block should the browser exit.

This is currently not implemented on the browser side, I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1903338 for that.

Is the session automatically canceled (and the prompt dismissed) if the session is closed (e.g. due to the browser shutting down)? Could we have a unit test for that?